exchanges/okgroup/okgroup_websocket.go

```diff
@@ -434,178 +559,344 @@ func (o *OKGroup) wsProcessCandles(response *WebsocketDataResponse) {
 	}
-	asks := o.AppendWsOrderbookItems(wsEventData.Asks)
-	bids := o.AppendWsOrderbookItems(wsEventData.Bids)
+
+	asks, err := o.AppendWsOrderbookItems(wsEventData.Asks)
+	if err != nil {
+		return err
+	}
+
+	bids, err := o.AppendWsOrderbookItems(wsEventData.Bids)
+	if err != nil {
+		return err
+	}
+
 	newOrderBook := orderbook.Base{
 		Asks:         asks,
 		Bids:         bids,
-		AssetType:    o.GetAssetTypeFromTableName(tableName),
+		AssetType:    a,
 		LastUpdated:  wsEventData.Timestamp,
 		Pair:         instrument,
 		ExchangeName: o.GetName(),
 	}
 
-	err := o.Websocket.Orderbook.LoadSnapshot(&newOrderBook)
+	err = o.Websocket.Orderbook.LoadSnapshot(&newOrderBook)
 	if err != nil {
 		return err
 	}
+
+	// This sets a valid checksumed and verified orderbook
+	o.LocalOB[wsEventData.InstrumentID] = newOrderBook
+
 	o.Websocket.DataHandler <- wshandler.WebsocketOrderbookUpdate{
 		Exchange: o.GetName(),
-		Asset:    o.GetAssetTypeFromTableName(tableName),
+		Asset:    a,
 		Pair:     instrument,
 	}
 	return nil
 }
 
 // WsProcessUpdateOrderbook updates an existing orderbook using websocket data
 // After merging WS data, it will sort, validate and finally update the existing orderbook
-func (o *OKGroup) WsProcessUpdateOrderbook(wsEventData *WebsocketDataWrapper, instrument currency.Pair, tableName string) error {
+func (o *OKGroup) WsProcessUpdateOrderbook(wsEventData *WebsocketDataWrapper, instrument currency.Pair, a asset.Item) error {
 	update := wsorderbook.WebsocketOrderbookUpdate{
-		AssetType:    asset.Spot,
+		AssetType:    a,
 		CurrencyPair: instrument,
 		UpdateTime:   wsEventData.Timestamp,
 	}
-	update.Asks = o.AppendWsOrderbookItems(wsEventData.Asks)
-	update.Bids = o.AppendWsOrderbookItems(wsEventData.Bids)
-	err := o.Websocket.Orderbook.Update(&update)
+
+	var err error
+	update.Asks, err = o.AppendWsOrderbookItems(wsEventData.Asks)
 	if err != nil {
-		log.Error(log.ExchangeSys, err)
+		return err
+	}
+	update.Bids, err = o.AppendWsOrderbookItems(wsEventData.Bids)
+	if err != nil {
+		return err
 	}
-	updatedOb := o.Websocket.Orderbook.GetOrderbook(instrument, asset.Spot)
-	checksum := o.CalculateUpdateOrderbookChecksum(updatedOb)
-	if checksum == wsEventData.Checksum {
-		if o.Verbose {
-			log.Debug(log.ExchangeSys, "Orderbook valid")
-		}
-		o.Websocket.DataHandler <- wshandler.WebsocketOrderbookUpdate{
-			Exchange: o.GetName(),
-			Asset:    o.GetAssetTypeFromTableName(tableName),
-			Pair:     instrument,
-		}
 
-	} else {
-		if o.Verbose {
-			log.Warnln(log.ExchangeSys, "Orderbook invalid")
-		}
-		return fmt.Errorf("channel: %v. Orderbook update for %v checksum invalid. Received %v Calculated %v", tableName, instrument, wsEventData.Checksum, checksum)
+	scopedOB, ok := o.LocalOB[wsEventData.InstrumentID]
+	if !ok {
+		return nil
+	}
+
+	if wsEventData.Timestamp.Before(scopedOB.LastUpdated) {
+		return nil
+	}
+
+	scopedOB.Asks = ProcessOB(scopedOB.Asks, update.Asks, true)
```

Comment:
Similarly, is ProcessOB func necessary anymore either? `o.Websocket.Orderbook.Update(&update)` per previous performs the same logic in handling orderbook updates and sorting them, then processing the update